src/unix/kqueue.c

```diff
@@ -52,12 +52,38 @@ static void uv__fs_event(uv_loop_t* loop, uv__io_t* w, unsigned int fflags);
 
 
 int uv__kqueue_init(uv_loop_t* loop) {
+  struct kevent ev[3];
+  struct timespec timeout = {1, 0};
+  (void) (ev);
+  (void) (timeout);
+
   loop->backend_fd = kqueue();
   if (loop->backend_fd == -1)
     return UV__ERR(errno);
 
   uv__cloexec(loop->backend_fd, 1);
 
+#if UV__KQUEUE_EVFILT_USER
+  /* When using EVFILT_USER event to wake up the kqueue, this event must be
+   * registered beforehand. Otherwise, calling kevent() to issue an
+   * unregistered EVFILT_USER event will get an ENOENT.
+   * Since uv__async_send() may happen before uv__io_poll() with multi-threads,
+   * we can't defer this registration of EVFILT_USER event as we did for other
+   * events, but must perform it right away. Therefore, we register an
+   * EVFILT_USER event right after the kqueue instance is created and perform
+   * the runtime detection to check if kqueue with EVFILT_USER actually works.
+   */
+  EV_SET(ev, UV__KQUEUE_EVFILT_USER_IDENT, EVFILT_USER, EV_ADD | EV_CLEAR, 0, 0, 0);
+  EV_SET(ev + 1, UV__KQUEUE_EVFILT_USER_IDENT, EVFILT_USER, 0, NOTE_TRIGGER, 0, 0);
```

Comment:
Why does it need to trigger the event? Wouldn't the first EV_ADD already fail with EINVAL?